### ISSUE TITLE
Ability to slightly delay camera feed display (ARKit/ARCore only)?

### ISSUE THREAD

**Issue by AndrewJDR** [immersive-web/webxr-ar-module#44]
Note, I had a brief email exchange with @blairmacintyre and he said it'd be best to open an issue here and in the computer-vision repo regarding this, along with a reference to the computer-vision issue:
https://github.com/immersive-web/computer-vision/issues/1

I'll also paste the issue contents below:
We've developed a 2D/3D asset management and inspection tool that uses server-side rendering. Rendered frames are encoded to a video stream and sent back to client devices for display. This allows for dense assets to be interactive even on lower end mobile devices. In addition to standard interaction models (touch/mouse/keyboard controls), we also have an AR viewing mode. You can see the AR functionality in action to visualize a 1 billion triangle mesh at our Siggraph Real Time Live Demo at the 6:30 mark:
https://youtu.be/BQM9WyrXie4?t=389

This is happening inside chrome canary on android. You'll see at the 6:56 mark that the mesh, while workable, does have some issues with being slightly behind-sync with the real world environment. The reason it is behind sync is that this approach (and essentially all such server-side rendering approaches) must send the view transform from client to server, render a frame, encode the frame into a video stream, send that back to the client, decode, and overlay it with the AR camera display. While this turnaround can be made fairly fast (a few tens of milliseconds), the 3D content will always be slightly behind the video.

If we put aside HMDs and focus on arkit/arcore, if we only had some means of controlling when frames from the camera are displayed on the user's screen, it would make it possible to introduce a slight delay and synchronize the display of camera frames with the rendered results from the server. Is there anything planned within webxr to allow for this? Thanks.

**Comment by AndrewJDR** [comment]
Just wanted to check in on this one. Has there been any discussion around it? It's still something that's needed, from our perspective...

**Comment by AndrewJDR** [comment]
I wanted to circle back once more on this as we continue to get demand for this capability from users of our systems.
I'm going to tag the primary contacts on the WebXR AR module as shown here:
https://github.com/w3ctag/design-reviews/issues/462
@Manishearth @NellWaliczek @toji @AdaRoseCannon @cwilso 

For newcomers to the thread that want to catch up quickly, I've prepared an example of what this new API surface could look like here:
https://gist.github.com/AndrewJDR/e20ff4db3cd2c0f2409acf66da5c915a

**Comment by Manishearth** [comment]
This spec is nearing CR and we're not going to be adding any major APIs to this spec itself any time soon. Any new work would probably need to be done in a separate module.  It's also not clear if, for this feature, there is much implementor interest.

You can propose incubations [here](https://github.com/immersive-web/proposals/) (I'd recommend closing this issue and opening one there). Ideally you get an implementor on board too.

**Comment by Manishearth** [comment]
Oh, misread which repo this was on, ignore the previous (now deleted) comment if you got an email.

I kinda feel like this is sufficiently complex to require a separate incubation (see the process [here](https://github.com/immersive-web/proposals/). This seems like a pretty major feature and it's not clear to me if there's implementor interest which is crucial to add something to an implemented spec.

The proposed API won't work: the WebXR framebuffer does not have control over the camera feed; that's composited later. You'd at least need access to camera frames [from the CV incubation](https://github.com/immersive-web/computer-vision), and you'd need an additional API that gave control over the composited camera frame.

I know there's some interest from Google's side on exposing raw camera frames in AR, but I'm not sure if they'd be interested in adding control over what camera frames get composited.

Perhaps you should open an issue on the proposals repo and see if you can get interest there?

**Comment by AndrewJDR** [comment]
@Manishearth To be clear, there's no theoretical reason I need actual access to the camera feed for this to work -- an opaque handle to it could be fine, coupled with the ability to control the timing of playout for a given handle. Does that make anything easier?

CV seems to be more about access to the actual frame's contents, which typically also starts to involve user permissions and approvals, which I was really hoping to avoid, with this. That is, ideally, there'd be no need to ask the user permission to their camera, since the application wouldn't gain  access to the actual framebuffer data.

**Comment by Manishearth** [comment]
@AndrewJDR right, but then you're talking about some new and strange GL capabilities, making it even less likely to belong in an existing spec

**Comment by AndrewJDR** [comment]
@Manishearth Okay, thanks. I think I'll try to formulate something for the proposals repo. I could probably be an implementor for Chromium if that would help bolster the case for it.

**Comment by Manishearth** [comment]
@AndrewJDR you misunderstand, when I say implementor i don't mean a particular person willing to write the code, I mean a particular browser willing to ship it -- it would be helpful (but not a prerequisite) if you can convince a browser to work with you and agree to ship it

**Comment by AndrewJDR** [comment]
@Manishearth Regarding implementor: Ah understood. I think I'll wait until we have our non-browser based implementation of this finalized before filing an issue in the proposals repo, so folks have a concrete usecase they can see in action. Thanks again.
